### PR TITLE
fixed #4511

### LINKED ISSUE #4511: AttachmentsClassLoader fails when calling setURLStreamHandlerFactory
* A clear description of the issue

`AttachmentsClassLoader` blindly attempts to set a custom `URLStreamHandlerFactory` by calling `URL.setURLStreamHandlerFactory`. The problem with this approach is the method can be [called at most once in a given Java Virtual Machine](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#setURLStreamHandlerFactory(java.net.URLStreamHandlerFactory)), so  
`AttachmentsClassLoader` fails in case a `URLStreamHandlerFactory` is already set by something else, e.g. by Tomcat during integration testing.

For JDK8 this can be solved with using a decorator and reflection ([example](https://stackoverflow.com/questions/41696088/register-custom-urlstreamhandler-in-spring-web-application-tomcat#answer-41760601)). As of Java 9 you can use SPI to provide handlers.

* Any logs or stack traces that you can provide (use a site like https://pastebin.com for larger logs)

 e.g. when TransactionBuilder.toSignedTransaction is called:

```
java.lang.Error: factory already defined
        at java.net.URL.setURLStreamHandlerFactory(URL.java:1112) ~[na:1.8.0_181]
        at net.corda.core.serialization.internal.AttachmentsClassLoader.<clinit>(AttachmentsClassLoader.kt:45) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.core.serialization.internal.AttachmentsClassLoaderBuilder$build$2.apply(AttachmentsClassLoader.kt:146) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.core.serialization.internal.AttachmentsClassLoaderBuilder$build$2.apply(AttachmentsClassLoader.kt:136) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at java.util.HashMap.computeIfAbsent(HashMap.java:1127) ~[na:1.8.0_181]
        at java.util.Collections$SynchronizedMap.computeIfAbsent(Collections.java:2672) ~[na:1.8.0_181]
        at net.corda.core.serialization.internal.AttachmentsClassLoaderBuilder.build(AttachmentsClassLoader.kt:145) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.core.serialization.internal.AttachmentsClassLoaderBuilder.withAttachmentsClassloaderContext(AttachmentsClassLoader.kt:153) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.core.transactions.LedgerTransaction.verifyInternal$core(LedgerTransaction.kt:135) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.core.transactions.LedgerTransaction.verify(LedgerTransaction.kt:121) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.core.transactions.TransactionBuilder.addMissingDependency(TransactionBuilder.kt:181) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.core.transactions.TransactionBuilder.toWireTransactionWithContext$core(TransactionBuilder.kt:168) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.core.transactions.TransactionBuilder.toWireTransactionWithContext$core$default(TransactionBuilder.kt:136) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.core.transactions.TransactionBuilder.toWireTransaction(TransactionBuilder.kt:133) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.core.transactions.TransactionBuilder.toSignedTransaction(TransactionBuilder.kt:701) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.core.node.ServiceHub$DefaultImpls.signInitialTransaction(ServiceHub.kt:228) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.core.node.ServiceHub$DefaultImpls.signInitialTransaction(ServiceHub.kt:242) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.node.services.api.ServiceHubInternal$DefaultImpls.signInitialTransaction(ServiceHubInternal.kt) ~[corda-node-4.0-SNAPSHOT.jar:na]
        at net.corda.node.internal.AbstractNode$ServiceHubInternalImpl.signInitialTransaction(AbstractNode.kt:994) ~[corda-node-4.0-SNAPSHOT.jar:na]
        at net.corda.core.node.ServiceHub$DefaultImpls.signInitialTransaction(ServiceHub.kt:251) ~[corda-core-4.0-SNAPSHOT.jar:na]
        at net.corda.node.services.api.ServiceHubInternal$DefaultImpls.signInitialTransaction(ServiceHubInternal.kt) ~[corda-node-4.0-SNAPSHOT.jar:na]
        at net.corda.node.internal.AbstractNode$ServiceHubInternalImpl.signInitialTransaction(AbstractNode.kt:994) ~[corda-node-4.0-SNAPSHOT.jar:na]
```

same error when calling TransactionBuilder.verify

```
java.lang.Error: factory already defined
	at java.net.URL.setURLStreamHandlerFactory(URL.java:1112) ~[na:1.8.0_181]
	at net.corda.core.serialization.internal.AttachmentsClassLoader.<clinit>(AttachmentsClassLoader.kt:45) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at net.corda.core.serialization.internal.AttachmentsClassLoaderBuilder$build$2.apply(AttachmentsClassLoader.kt:146) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at net.corda.core.serialization.internal.AttachmentsClassLoaderBuilder$build$2.apply(AttachmentsClassLoader.kt:136) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at java.util.HashMap.computeIfAbsent(HashMap.java:1127) ~[na:1.8.0_181]
	at java.util.Collections$SynchronizedMap.computeIfAbsent(Collections.java:2672) ~[na:1.8.0_181]
	at net.corda.core.serialization.internal.AttachmentsClassLoaderBuilder.build(AttachmentsClassLoader.kt:145) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at net.corda.core.serialization.internal.AttachmentsClassLoaderBuilder.withAttachmentsClassloaderContext(AttachmentsClassLoader.kt:153) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at net.corda.core.transactions.LedgerTransaction.verifyInternal$core(LedgerTransaction.kt:135) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at net.corda.core.transactions.LedgerTransaction.verify(LedgerTransaction.kt:121) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at net.corda.core.transactions.TransactionBuilder.addMissingDependency(TransactionBuilder.kt:181) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at net.corda.core.transactions.TransactionBuilder.toWireTransactionWithContext$core(TransactionBuilder.kt:168) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at net.corda.core.transactions.TransactionBuilder.toWireTransactionWithContext$core$default(TransactionBuilder.kt:136) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at net.corda.core.transactions.TransactionBuilder.toWireTransaction(TransactionBuilder.kt:133) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at net.corda.core.transactions.TransactionBuilder.toLedgerTransaction(TransactionBuilder.kt:483) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at net.corda.core.transactions.TransactionBuilder.verify(TransactionBuilder.kt:491) ~[corda-core-4.0-SNAPSHOT.jar:na]
	at mypackage.yo.workflow.YoFlow.call(YoWorkflow.kt:73) ~[cordapp-workflow-0.19-SNAPSHOT.jar:na]
	at mypackage.yo.workflow.YoFlow.call(YoWorkflow.kt:41) ~[cordapp-workflow-0.19-SNAPSHOT.jar:na]
```

* The version/tag/release or commit hash it occurred on

master

### PR DESCRIPTION
**Pull request by manosbatsis**
This PR uses a decorator and reflection to apply `AttachmentURLStreamHandlerFactory` when `URL.setURLStreamHandlerFactory` has already been called. 

This seems to be the only solution for JDK8 that keeps resource resolution contained within `URL`. A quick search suggests that, starting with Java 9, either adding `--add-opens` or going through an extension/service loader will be needed.

I state that this PR is in accordance with the [Developer's Certificate of Origin](https://github.com/corda/corda/blob/master/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by manosbatsis** [comment]
@joeldudleyr3 @tudor-malene https://github.com/corda/corda/issues/4511 was breaking all my integration tests with master so thought i might as well help myself. Could you please push this through or let me know if anything is needed?

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt [hunk 1]
```diff
@@ -41,10 +43,44 @@ class AttachmentsClassLoader(attachments: List<Attachment>, parent: ClassLoader
         private val log = contextLogger()
 
         init {
-            // This is required to register the AttachmentURLStreamHandlerFactory.
-            URL.setURLStreamHandlerFactory(AttachmentURLStreamHandlerFactory)
+            // Retrieve the `URL.factory` field
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: Can you move this logic into a method called: `setOrDecorateURLStreamHandlerFactory`, as it quite large.

Also can you log a warning on the else branch: 
`log.warn("The URLStreamHandlerFactory was already set in the JVM. Please be aware that this is not recommended.")`

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt [hunk 2]
```diff
@@ -167,7 +203,7 @@ internal object AttachmentsClassLoaderBuilder {
  * This will not be exposed as an API.
  */
 object AttachmentURLStreamHandlerFactory : URLStreamHandlerFactory {
-    private const val attachmentScheme = "attachment"
+    internal const val attachmentScheme = "attachment"
```
Comment (tudor-malene) [inline core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: Does this need to be internal?

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt (hunk 1 above)
Comment (manosbatsis) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: Done!

File: core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt (hunk 2 above)
Comment (manosbatsis) [reply core/src/main/kotlin/net/corda/core/serialization/internal/AttachmentsClassLoader.kt]: Nope, don't even recall setting as such. Good one.

**Comment by tudor-malene** [comment]
Hi Manos, thank you for the changes.

Can you please merge in the latest master, so we get the green build, and then we can merge!

**Comment by manosbatsis** [comment]
@tudor-malene Looks to me `AttachmentsClassLoader` is my only diff with upstream master but teamcity "Integration Tests (Pull Requests)" fail, possibly due to port conflicts. Perhaps i'm missing something, anything i can do?

**Comment by joeldudleyr3** [comment]
@manosbatsis Hey Manos, the code looks perfect but unfortunately it looks like you've accidentally included some other commits in your branch (e.g. check #4478 here: https://github.com/corda/corda/pull/4512/commits).

Can you create a fresh branch off Corda's `master` branch, reapply your changes and then re-open the PR? We can then merge the changes in. Let me know if you run into issues :)

**Pull request merged**